src/decoder/mod.rs

```diff
@@ -196,16 +188,25 @@ fn decode_bson<R: Read + ?Sized>(reader: &mut R, tag: u8, utf8_lossy: bool) -> D
         Some(ElementType::EmbeddedDocument) => decode_document(reader).map(Bson::Document),
         Some(ElementType::Array) => decode_array(reader, utf8_lossy).map(Bson::Array),
         Some(ElementType::Binary) => {
-            let len = read_i32(reader)?;
+            let mut len = read_i32(reader)?;
```

Comment:
The tests check that we automatically handle the repeated length field in binary subtype 2, so I updated the logic here to do that